### DISCUSSION TITLE
关于自定义组件数据的监听和事件联动

### DISCUSSION THREAD

**Discussion by jianxiaoBai** [alibaba/formily#2690] [Q&A 2.x] [answered]
背景：
目前我们是在用 formily 用 schema 的方式实现 CURD 的操作，在 CURD 中经常会有 query 和 table 和 pagination 的互相联动，需要组件数据之间的监听，比如分页变化是需要发起请求更新 table 数据，查看官方文档后发现有对应 `onFieldReact` API有对应的功能，因为父子组件的关系需要提前定义好，耦合严重，好像也无法在 schema 模式下应用。

方案探索：
每个组件都通过 model 定义自己的数据模型，然后该模型的实例是放到全局的 Map 数据结构里面，key 是 组件 path， value 是 数据模型的实例，这样的话其他组件就可以只通过路径就可以拿到 model 实例，进行数据监听和事件的触发，model 的数据和组件的数据是一致的，交互式大概就是

当 A 组件监听 B 组件时的交互流程：

A 组件 A model
B 组件 B model

A 组件中依赖 A model 的值然后通过 field.setValue 更新自身，
A model 又监听了 B model 的 value，这样当 B 组件事件触发时会更新自己的 B model，从而监听 B model 的 A model 也会更新自身，从而 A 组件也会变化。

这样来实现组件间的数据监听同时也可以实现跨组件间的时间触发。

请教一下这样的方案可行吗，或者有没有更好的方案实现呢？

下面这张图，是自定义 Object 组件的 model 的实例化和代码关系。
![image](https://user-images.githubusercontent.com/20953838/147525971-cef40fcf-95f9-4e13-a21b-a18e97a2ed0c.png)

**Comment by janryWang** [comment, marked as answer]
你这个数据模型是干嘛用的，为啥不直接用Field模型

**Reply by jianxiaoBai** [reply]
我愚了，field 的数据模型是在 reactive 定义的，能通过路径获取到了自然也可以监听对吧，我当时在群里问过这样的问题，不过当时没有答复，我就自己想办法，就走歪了。

**Reply by janryWang** [reply]
嗯嗯，以后遇到一句话讲不清楚的问题，就开个discussion，不然群里也很难达到信息绝对对称

**Reply by cgfeel** [reply]
如果自定义按钮呢，要监听数据变化，没有数据的时候 `disabled`
```
                    <FormButtonGroup>
                        <Submit onSubmit={console.log}>提交</Submit>
                        <FormBtn
                            onClick={() => {
                                form.setInitialValues({
                                    table_list: range(100000),
                                });
                            }}
                            block>
                            加载10W条超大数据
                        </FormBtn>
                    </FormButtonGroup>
```
